controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -92,20 +129,317 @@ func (r *StorageClassClaimReconciler) Reconcile(ctx context.Context, request rec
+		pvList := corev1.PersistentVolumeList{}
+		err := r.list(&pvList, client.MatchingFields{
+			"spec.storageClassName": r.storageClassClaim.Name,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to list PersistentVolumes: %s", err)
+		}
+		if len(pvList.Items) > 0 {
+			return fmt.Errorf("StorageClass %s is still in use by one or more PV(s)", r.storageClassClaim.Name)
+		}
+
+		// Remove finalizer from StorageClass.
+		storageClass.SetFinalizers(remove(storageClass.Finalizers, StorageClassFinalizer))
+		if err := r.update(storageClass); err != nil {
+			return fmt.Errorf("failed to remove finalizer from StorageClass: %s", err)
+		}
+
+		// Delete the StorageClass.
+		err = r.delete(existing)
+		if err != nil {
+			r.log.Error(err, "Failed to delete StorageClass.", "StorageClass", klog.KRef(storageClass.Namespace, existing.Name))
+			return err
+		}
+	}
+	r.log.Info("Creating StorageClass.", "StorageClass", klog.KRef(storageClass.Namespace, existing.Name))
+	err := r.Client.Create(r.ctx, storageClass)
+	if err != nil {
+		return fmt.Errorf("failed to create StorageClass: %v", err)
+	}
+	return nil
+}
+
+func (r *StorageClassClaimReconciler) reconcileProviderPhases() (reconcile.Result, error) {
+	log := ctrllog.FromContext(r.ctx)
 	log.Info("Running StorageClassClaim controller in Converged/Provider Mode")
 	return reconcile.Result{}, nil
 }
 
 func (r *StorageClassClaimReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&ocsv1alpha1.StorageClassClaim{}).
+		For(&ocsv1alpha1.StorageClassClaim{}, builder.WithPredicates(predicate.Funcs{
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				// Validation check.
+				oldStorageClassClaim := e.ObjectOld.(*ocsv1alpha1.StorageClassClaim)
+				newStorageClassClaim := e.ObjectNew.(*ocsv1alpha1.StorageClassClaim)
+				newStorageClassClaimKRef := klog.KRef(newStorageClassClaim.Namespace, newStorageClassClaim.Name)
+				switch {
+				case oldStorageClassClaim.Spec.Type != newStorageClassClaim.Spec.Type:
+					r.log.Info("StorageClassClaim type change not allowed, skipping update",
+						"StorageClassClaim", newStorageClassClaimKRef)
+					return false
+				case oldStorageClassClaim.Spec.EncryptionMethod != newStorageClassClaim.Spec.EncryptionMethod:
+					r.log.Info(
+						"StorageClassClaim encryption method change not allowed, skipping update",
+						"StorageClassClaim", newStorageClassClaimKRef)
+					return false
+				default:
+					return true
+				}
+			},
+		})).
+		Owns(&storagev1.StorageClass{}).
 		Complete(r)
 }
+
+func (r *StorageClassClaimReconciler) get(obj client.Object) error {
+	key := client.ObjectKeyFromObject(obj)
+	return r.Client.Get(r.ctx, key, obj)
+}
+
+func (r *StorageClassClaimReconciler) list(obj client.ObjectList, listOptions ...client.ListOption) error {
+	if len(listOptions) == 0 {
+		return r.Client.List(r.ctx, obj, client.InNamespace(r.namespace))
+	}
+	return r.Client.List(r.ctx, obj, listOptions[0])
```

Comment:
why only listOptions[0], and not full list?

Reply:
This stemmed from a desire to have dynamic function signature for `list`. If the caller does not want to supply any explicit `listOptions`, they don't need to pass `null` to indicate that. Instead, due to the variadic function argument we can determine if a `1istOption` parameter was supplied and pass that on to `List` if so.

Reply:
I think we are are doing that with the `len(listOptions) == 0`

My question is when someone passes the list option and the length is not 0 why are we sending only the first list option and not the whole list?

Reply:
Okay, let's assume the function is called like so: `list(obj, listOptionsA, listOptionsB, ..., listOptionsZ)`. The variadic nature of the function does not prevent it from accepting more than one `listOption`, but allows the function to be called as `list(obj)` as well. Now practically, if a custom `list` function replaces a single `List` call, then the function itself is expected to make a single `List` call internally (and the fact that a single `List` call does not take in an array of `listOptions`). So right now, only `listOptionsA` will be applied to `obj`.

PLMK if you think the variadic argument is counter-intuitive here and I'll refactor the method to a more simpler outlook, thanks.

Reply:
Got it after trying this out, I think we can let it be like that only